numpy/core/src/multiarray/common.c

```diff
@@ -133,7 +133,7 @@ NPY_NO_EXPORT PyArray_Descr *
 _array_typedescr_fromstr(char const *c_str)
 {
     PyArray_Descr *descr = NULL;
-    PyObject *stringobj = PyString_FromString(c_str);
+    PyObject *stringobj = PyBytes_FromString(c_str);
```

Comment:
This should be `PyUnicode_FromString`, otherwise it is a (likely  incorrect) behavior change.

Reply:
Hmm... maybe. This doesn't change current behavior as this is the current definition of `PyString_FromString` for Python 3, but it may only be working by accident or not tested. @seberg Thoughts?

Reply:
Hmm, I'm confused them. This patch is odd, because it replaces `PyString_FromString` with `PyUnicode_FromString` in some places, but `PyBytes_FromString` in others.

I didn't actually check which was which, but I assumed this one as the minority was the mistake.

Reply:
Hmmm... This function has only two callers:

https://github.com/numpy/numpy/blob/7ec2e1bac72afcdc68cf8256879afbc4cb14a907/numpy/core/src/multiarray/ctors.c#L1779-L1781

https://github.com/numpy/numpy/blob/7ec2e1bac72afcdc68cf8256879afbc4cb14a907/numpy/core/src/multiarray/ctors.c#L1892-L1910

The first clearly doesn't care about string vs unicode (and the latter would be faster, because `PyArray_DescrConverter` starts by converting `str` to unicode).

The second seems a bit silly, as we take a unicode object, encode it to bytes, convert to char*, and then do the whole lot in reverse again.

Basically, this change is fine and I was wrong to think it a behavior change - but there's more cleanup that could be done here.

Reply:
> This patch is odd, because it replaces PyString_FromString with PyUnicode_FromString

Yes. The other places looked like either would work and I thought unicode would be the better choice now that Python 2.7 support is gone.